baseapp/abci.go

```diff
@@ -561,6 +561,59 @@ func (app *BaseApp) createQueryContext(height int64, prove bool) (sdk.Context, e
 	return ctx, nil
 }
 
+// GetBlockRentionHeight returns the height for which all blocks below this height
+// are pruned from Tendermint. Given a commitment height and a non-zero local
+// minRetainBlocks configuration, the retentionHeight is the smallest height that
+// satisfies:
+//
+// - Unbonding (safety threshold) time: The block interval in which validators
+// can be economically punished for misbehavior. Blocks in this interval must be
+// auditable e.g. by the light client.
+//
+// - Logical store snapshot interval: The block interval at which the underlying
+// logical store database is persisted to disk, e.g. every 10000 heights. Blocks
+// since the last IAVL snapshot must be available for replay on application restart.
+//
+// - State sync snapshots: Blocks since the oldest available snapshot must be
+// available for state sync nodes to catch up (oldest because a node may be
+// restoring an old snapshot while a new snapshot was taken).
+//
+// - Local (minRetainBlocks) config: Archive nodes may want to retain more or
+// all blocks, e.g. via a local config option min-retain-blocks. There may also
+// be a need to vary retention for other nodes, e.g. sentry nodes which do not
+// need historical blocks.
+func (app *BaseApp) GetBlockRentionHeight(commitHeight uint64) uint64 {
```

Comment:
@erikgrinaker lmk if this matches what you have defined the [spec](https://github.com/tendermint/spec/blob/master/rfc/001-block-retention.md)

Reply:
Yes, the above comment matches the outline in the RFC. Note that this was just a suggestion, and there may be other factors that need to be taken into account - although I can't think of any.

There are a couple of issues with the logic below though:

- Shouldn't we check `commitHeight - app.minRetainBlocks`? I'm assuming `minRetainBlocks` is the number of recent blocks to keep, not the height to keep.

- If the user disabled state sync snapshotting by setting `snapshotInterval = 0`, this will cause block pruning to be disabled (since min(0, x) is always 0). The same goes for state pruning and evidence age. These constraints should probably only be applied if they are non-zero.

- I don't know if evidence age is the right metric to use, as opposed to the validator unbonding time - I'm not sure how these are related, and how they can be controlled.

Reply:
Good point on the first two. WRT to the last point, max evidence age is the unbonding period, in other words, the max evidence age == block time * unbonding period days.